libs/x-content/src/test/java/org/elasticsearch/xcontent/XContentParserTests.java

```diff
@@ -475,6 +475,37 @@ public void testSubParserObject() throws IOException {
         }
     }
 
+    public void testFlatteringParserObject() throws IOException {
+        String content = """
+            {
+              "parent": {
+                "child1" : 1,
+                "child2": {
+                  "grandChild" : 1
+                }
+              }
+            }
+            """;
+        XContentParser parser = createParser(JsonXContent.jsonXContent, content);
+        assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+        assertEquals(XContentParser.Token.FIELD_NAME, parser.nextToken());
+        assertEquals("parent", parser.currentName());
+        assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+        XContentParser subParser = new FlatteringXContentParser(parser, parser.currentName());
+        assertEquals(XContentParser.Token.FIELD_NAME, subParser.nextToken());
+        assertEquals("parent.child1", subParser.currentName());
+        assertEquals(XContentParser.Token.VALUE_NUMBER, subParser.nextToken());
+        assertEquals(XContentParser.Token.FIELD_NAME, subParser.nextToken());
+        String secondChildName = subParser.currentName();
+        assertEquals("parent.child2", secondChildName);
+        assertEquals(XContentParser.Token.START_OBJECT, subParser.nextToken());
+        XContentParser subSubParser = new FlatteringXContentParser(subParser, secondChildName);
```

Comment:
I agree on the `child3` test and I added it, but I think re-wrapping is necessary since `FlatteringXContentParser::currentName` handles only immediate child, therefore in the case you mentioned we'd get just `grandchild` as currentName plus in the code we'd re-wrap in case we encounter an object.

Reply:
> Rather than rewrapping here, can we test that we get `grandchild` as the next field name?
Sorry I read the comment wrongly. Now we check that we get `grandchild` as the next field name